### PR TITLE
✨ Store query MRNs with reporting jobs

### PR DESCRIPTION
**Pull request by czunker**
Add all MRNs which resolv to the same code id with their reporting job.

### CODE REVIEW

**Comment by czunker** [comment]
This would supersede https://github.com/mondoohq/cnspec/pull/1135

File: policy/resolver.go [hunk 1]
```diff
@@ -1163,31 +1178,44 @@ func (cache *policyResolverCache) addCheckJobVariants(ctx context.Context, query
...
 		cache.global.codeIdToMrn[query.CodeId] = append(cache.global.codeIdToMrn[query.CodeId], query.Mrn)
-		cache.global.reportingJobsByUUID[uuid] = queryJob
-		cache.global.reportingJobsByMsum[query.Checksum] = append(cache.global.reportingJobsByMsum[query.Checksum], queryJob)
-		cache.childJobsByMrn[query.Mrn] = append(cache.childJobsByMrn[query.Mrn], queryJob)
+		cache.global.reportingJobsByUUID[uuid] = rj
+		cache.global.reportingJobsByMsum[query.Checksum] = append(cache.global.reportingJobsByMsum[query.Checksum], rj)
+		cache.childJobsByMrn[query.Mrn] = append(cache.childJobsByMrn[query.Mrn], rj)
+	}
+
+	if _, ok := cache.global.reportingJobsByCodeId[query.CodeId]; !ok {
```
Comment (czunker) [inline policy/resolver.go:1186]: I'm not sure we also want to add it here.
Current focus might be only on checks.

File: policy/resolver_test.go [hunk 2]
```diff
@@ -1451,3 +1453,54 @@ queries:
...
 	})
 }
+
+func TestResolve_TwoMrns(t *testing.T) {
+	b := parseBundle(t, `
+owner_mrn: //test.sth
+policies:
+- uid: policy1
+  groups:
+  - type: chapter
```
Comment (jaym) [inline policy/resolver_test.go]: Can we add one more check where the code id matches but the filter does not match the asset

File: policy/resolver_test.go [hunk 3]
```diff
@@ -1451,3 +1453,54 @@ queries:
...
+      mql: asset.name == props.name
+      props:
+      - uid: name
+        mql: return "definitely not the asset name"
+    - uid: check2
+      mql: asset.name == props.name
+      props:
+      - uid: name
+        mql: return "definitely not the asset name"
+`)
```
Comment (jaym) [inline policy/resolver_test.go:1536]: it might be good to also have a queries example as the data queries have their reporting jobs built differently

**Comment by jaym** [comment]
Do we need special handling for variants? I think I would expect that the specific query that gets executed for a variant would end up with the variant's mrn. Right now, I think these have their mrn reported in their code id field since they have not code

File: internal/bundle/bundle.yac.go [hunk 4]
```diff
@@ -368,14 +368,15 @@ func (d ControlRef) MarshalYAML() (interface{}, error) {
...
-	Checks      []*Mquery   `protobuf:"bytes,21,rep,name=checks,proto3" json:"checks,omitempty" yaml:"checks,omitempty"`
-	Queries     []*Mquery   `protobuf:"bytes,22,rep,name=queries,proto3" json:"queries,omitempty" yaml:"queries,omitempty"`
-	FileContext FileContext `json:"-" yaml:"-"`
-	Comments    Comments    `json:"-" yaml:"-"`
+	Uid         string        `protobuf:"bytes,3,opt,name=uid,proto3" json:"uid,omitempty" yaml:"uid,omitempty"`
+	Mrn         string        `protobuf:"bytes,4,opt,name=mrn,proto3" json:"mrn,omitempty" yaml:"mrn,omitempty"`
+	Desc        string        `protobuf:"bytes,2,opt,name=desc,proto3" json:"desc,omitempty" yaml:"desc,omitempty"`
+	Title       string        `protobuf:"bytes,1,opt,name=title,proto3" json:"title,omitempty" yaml:"title,omitempty"`
+	Checks      []*Mquery     `protobuf:"bytes,21,rep,name=checks,proto3" json:"checks,omitempty" yaml:"checks,omitempty"`
+	Controls    []*ControlRef `protobuf:"bytes,23,rep,name=controls,proto3" json:"controls,omitempty" yaml:"controls,omitempty"`
```
Comment (czunker) [inline internal/bundle/bundle.yac.go:376]: I'm not sure why this is part of the PR.
I didn't change that. Perhaps `policy/generate` was missing on a previous PR.

File: policy/resolver_test.go (hunk 2 above)
Comment (czunker) [reply policy/resolver_test.go]: Added.

File: policy/resolver_test.go (hunk 3 above)
Comment (czunker) [reply policy/resolver_test.go:1536]: Added.

**Comment by czunker** [comment]
> Do we need special handling for variants? I think I would expect that the specific query that gets executed for a variant would end up with the variant's mrn. Right now, I think these have their mrn reported in their code id field since they have no code

I added a test for variants:
https://github.com/mondoohq/cnspec/pull/1344/files#diff-7ee0cbfbc4c0c6d335de85d19eb6f7e77aa8e24b223323392c43d4e7943dd696R1608

I'm not sure we need special handöing for this. This PR does not change the value of qr_id. It just adds the MRNs in any case.

File: policy/cnspec_policy.proto [hunk 5]
```diff
@@ -525,6 +525,9 @@ message ReportingJob {
   map<string,bool> datapoints = 7;
   map<string,cnquery.explorer.Impact> child_jobs = 9;
   Type type = 10;
+  // We will only store a ReportingJob by codeID
+  // But we add all the MRNs to it, so we can easily retrieve the MRN for a check
```
Comment (arlimus) [inline policy/cnspec_policy.proto]: Love that you are adding comments! let's make sure they are more targeted to the field, eg:
"MRNs of all checks / queries that this reporting job represents"

File: policy/resolver.go [hunk 6]
```diff
@@ -376,11 +376,12 @@ type resolverCache struct {
...
 	propsCache       explorer.PropsCache
 
-	reportingJobsByUUID map[string]*ReportingJob
-	reportingJobsByMsum map[string][]*ReportingJob // Msum == Mquery.Checksum, i.e. only reporting jobs for mqueries
-	reportingJobsActive map[string]bool
-	errors              []*policyResolutionError
-	bundleMap           *PolicyBundleMap
+	reportingJobsByUUID   map[string]*ReportingJob
+	reportingJobsByMsum   map[string][]*ReportingJob // Msum == Mquery.Checksum, i.e. only reporting jobs for mqueries
+	reportingJobsByCodeId map[string][]*ReportingJob // CodeId == check.CodeId
```
Comment (arlimus) [inline policy/resolver.go]: replace `check` in the comment with either `Mquery` (to be consistent with the above) or query. ReportingJobs don't mind if it's a check or a data query

File: policy/resolver.go (hunk 1 above)
Comment (arlimus) [reply policy/resolver.go:1186]: I think it should be fine, but please keep an eye open for side-effects (and if necessary fix this)

**Review by arlimus — APPROVED** [review]
I added a few small things to fix, feel free to get those in and then merge. The rest looks great :+1:  Thank you @czunker :)

File: policy/resolver.go (hunk 6 above)
Comment (czunker) [reply policy/resolver.go]: Thanks for the feedback, changed this.

File: policy/cnspec_policy.proto (hunk 5 above)
Comment (czunker) [reply policy/cnspec_policy.proto]: Thanks for the feedback, changed this.

**Pull request merged**